### PR TITLE
Memory range introspection

### LINKED ISSUE #47: Alternative ways to compute a machine's state hash
## Context

We need to make sure the machine Merkle tree implementation works correctly and indeed captures the entire state of the machine. It is therefore important that we be able to reproduce the root hash without using the code in the machine itself. 

## Possible solutions

At least two ways of doing this would be useful.

We should be able to, using the machine interface, iterate over the contents of the machine and externally compute the state hash. This would be useful in tests.

We should be able to, without instantiating a machine, and iterating over the contents of a machine config, compute the state hash. This would be useful to check integrity of a stored machine directory.

## Subtasks

- [ ] Ensure the machine API exposes enough information to allow the external computation of the state hash
- [ ] Implement a function that computes the state hash given a machine instance
- [ ] Implement a function  that computes the state hash given a machine config

### PR DESCRIPTION
**Pull request by diegonehab**
This closes #47 and #96. It depends on grpc-interfaces PR [#24](https://github.com/cartesi/grpc-interfaces/pull/24)

### CODE REVIEW

File: src/tests/util.lua [hunk 1]
```diff
@@ -219,10 +219,8 @@ end
 function test_util.align(v, el) return (v >> el << el) end
 
 function test_util.load_file(filename)
-    local fd = assert(io.open(filename, "rb"))
-    local data_size = fd:seek("end")
-    fd:seek("set")
-    local data = fd:read(data_size)
+    local fd <close> = assert(io.open(filename, "rb"))
+    local data = assert(fd:read("*all"))
     fd:close(filename)
```
Comment (edubart) [inline src/tests/util.lua:224]: Unnecessary close

File: src/tests/machine-test.lua [hunk 2]
```diff
@@ -277,63 +237,44 @@ test_util.make_do_test(build_uarch_machine, machine_type)(
...
-
-    local ram_address_start = tonumber(test_util.split_string(ram_file_name, "--.")[1], 16)
-    local ram_log2_size = math.ceil(math.log(#ram, 2))
-    local calculated_ram_hash = test_util.merkle_hash(ram, 0, ram_log2_size)
-
-    local ram_proof = machine:get_proof(ram_address_start, ram_log2_size)
+    -- find ram memory range
+    local ram
+    for _, v in ipairs(machine:get_memory_ranges()) do
+        print(v.description)
```
Comment (edubart) [inline src/tests/machine-test.lua:263]: Unnecessary print

File: src/tests/machine-bind.lua [hunk 3]
```diff
@@ -545,7 +541,7 @@ end)
 print("\n\n dump pmas to files")
 do_test("there should exist dumped files of expected size", function(machine)
     -- Dump pmas to files
-    machine:dump_pmas()
+    machine:dump_memory_ranges()
 
     for i = 1, #pmas_file_names do
```
Comment (edubart) [inline src/tests/machine-bind.lua:546]: This hardcoded list of pmas could be moved inside this test. Morever the `pmas_sizes` could be extracted from `pmas_file_names`. Or the pmas could be listed from `get_memory_ranges`.

**Comment by diegonehab** [comment]
I'll just get rid of the method everywhere.

**Review by edubart — APPROVED** [review]

**Review by edubart — DISMISSED** [review]

**Review by edubart — DISMISSED** [review]

**Review by edubart — APPROVED** [review]

**Review by vfusco — APPROVED** [review]

**Review by edubart — APPROVED** [review]

**Pull request merged**